File: wrappers/python/memhole.py
# Python wrapper v1.0 for Memhole v1.2.x
# This file allows for easy use of memhole in any python program
# The memhole kernel module must be loaded for these functions to work properly

# note: unlike the v1.6+ C/C++ wrapper, this wrapper has no option
# for thread safety between io operations; it is not thread safe

import os

MEMHOLE_PATH = "/dev/memhole"

#error codes
EINVDEV = 4
EMEMHNF = 8
EMEMBSY = 16
EINVPID = 32
EKMALOC = 64


class Memhole():
    buf_size = 0
    write_perms = False
    fd = 0

    __LSMSPID = 0
    __LSMSPOS = 1
    __LSMGPOS = 2
    __LSMSBUF = 3

    def __init__(self, write_perms : bool):
        self.mode = mode
        self.write_perms = write_perms
    
    #returns 0 on success or an error code
    def connect(self) -> int:
        if self.fd != 0:
            return -EINVDEV
        mode = os.O_RDONLY
        if self.write_perms:
            mode = os.O_RDWR
        if not os.access(MEMHOLE_PATH, os.F_OK):
            return -EMEMHNF
        self.fd = os.open(MEMHOLE_PATH, mode)
        if self.fd < 1:
            self.fd = 0
            return -EMEMBSY
        self.buf_size = 0
        return 0
    
    #returns 0 on success or an error code
    def disconnect(self) -> int:
        if self.fd == 0:
            return -EINVDEV
        ret_val = os.close(self.fd)
        if ret_val != 0:
            return ret_val
        self.fd = 0
        return 0
    
    #returns 0 on success or an error code
    def attach_to_pid(self, pid):
        if self.fd == 0: return -EINVDEV
        if os.kill(pid, 0): return -EINVPID

        ret = os.lseek(self.fd, pid, self.__LSMSPID)
        if ret == 0: return -EINVPID
        return 0
    
    #returns 0 on success or an error code
    def set_memory_position(self, pos):
        if self.fd == 0: return -EINVDEV
        return os.lseek(self.fd, pos, self.__LSMSPOS)

    #returns current memory address on success or an error code
    def get_memory_position(self):
        if self.fd == 0: return -EINVDEV
        return os.lseek(self.fd, 0, self.__LSMGPOS)
    
    #returns 0 on success or an error code
    def set_buffer_size(self, len):
        if self.fd == 0: return -EINVDEV
        if(len < self.buf_size):
            if(os.lseek(self.fd, len, self.__LSMSBUF) != 0):
                return -EKMALOC
        return 0

    #returns a bytestring
    def read_memory(self, len):
        if self.fd == 0: return -EINVDEV
        if(self.set_buffer_size(len) != 0): return -EKMALOC
        return os.read(self.fd, len)
    
    #returns the number of bytes written
    def write_memory(self, buf_str, len):
        if self.fd == 0: return -EINVDEV
        if(self.set_buffer_size(len) != 0): return -EKMALOC
        return os.write(self.fd, buf_str)

File: wrappers/c#/Memhole.cs
using System.Diagnostics;

/*
 * Memhole C# Wrapper v1.0 for Memhole v1.2.x
 * This library allows for easy use of the Memhole kernel module for Linux
 * 
 * Note: unlike the v1.6+ C/C++ wrapper, this wrapper is not thread safe
 * You must implement thread safety on your own if you plan to access memhole
 * from multiple threads. (The memhole device itself is always locked to 1 instance)
 */

namespace Memhole
{
    public enum ErrorCodes : int
    {
        InvalidDevice = 4,
        MemholeDeviceNotFound = 8,
        MemholeDeviceBusy = 16,
        InvalidPID = 32,
        MemholeDeviceKmallocFail = 64
    };

    public class Memhole
    {
        private const string MEMHOLE_DEVICE_PATH = "/dev/memhole";

        private const int SEEK_MODE_SET_PID = 0;
        private const int SEEK_MODE_SET_POS = 1;
        private const int SEEK_MODE_GET_POS = 2;
        private const int SEEK_MODE_SET_BUF = 3;

        long bufferSize = 0;
        bool canWrite = false;
        FileStream? device = null;
        
        public Memhole(bool write = false)
        {
            canWrite = write;
        }

        public int Connect()
        {
            if(device != null)
            {
                return -((int) ErrorCodes.InvalidDevice);
            }

            if (!File.Exists(MEMHOLE_DEVICE_PATH))
            {
                return -((int)ErrorCodes.MemholeDeviceNotFound);
            }

            try
            {
                if (canWrite)
                {
                    device = File.Open(MEMHOLE_DEVICE_PATH, FileMode.Open, FileAccess.ReadWrite);
                }
                else
                {
                    device = File.Open(MEMHOLE_DEVICE_PATH, FileMode.Open, FileAccess.Read);
                }
            }
            catch
            {
                return -((int)ErrorCodes.MemholeDeviceBusy);
            }

            bufferSize = 0;
            return 0;
        }

        public int Disconnect()
        {
            if (device == null)
            {
                return -((int)ErrorCodes.InvalidDevice);
            }

            try
            {
                device.Close();
            }
            catch
            {
                return -((int)ErrorCodes.InvalidDevice);
            }

            device = null;
            return 0;
        }

        public int AttachToPid(int pid)
        {
            if (device == null)
            {
                return -((int)ErrorCodes.InvalidDevice);
            }
            if(!Process.GetProcesses().Any(x => x.Id == pid))
            {
                return -((int) ErrorCodes.InvalidPID);
            }

            long ret = device.Seek((long)pid, SEEK_MODE_SET_PID);
            return ret == 0 ? -((int) ErrorCodes.InvalidPID) : (int) ret;
        }

        public int SetMemoryPosition(long pos)
        {
            if (device == null)
            {
                return -((int)ErrorCodes.InvalidDevice);
            }

            return (int) device.Seek((long)pos, (SeekOrigin)(SEEK_MODE_SET_POS));
        }

        public int GetMemoryPosition()
        {
            if (device == null)
            {
                return -((int)ErrorCodes.InvalidDevice);
            }

            return (int)device.Seek((long)0, (SeekOrigin)(SEEK_MODE_GET_POS));
        }

        public int SetBufferSize(long len)
        {
            if (device == null)
            {
                return -((int)ErrorCodes.InvalidDevice);
            }

            if(len < bufferSize)
            {
                if(device.Seek(len, (SeekOrigin)(SEEK_MODE_SET_BUF)) != 0)
                {
                    return -((int)ErrorCodes.MemholeDeviceKmallocFail);
                }
            }

            return 0;
        }

        public byte[]? ReadMemory(int len)
        {
            if (device == null)
            {
                return null;
            }

            if(bufferSize < len)
            {
                if(SetBufferSize(len) != 0)
                {
                    return null;
                }
            }

            byte[] result = new byte[len];
            int bytesRead = device.Read(result, 0, len);
            if(bytesRead < result.Length)
            {
                Array.Resize(ref result, bytesRead);
            }
            return result;
        }

        public int WriteMemory(byte[] data)
        {
            if (device == null)
            {
                return -((int)ErrorCodes.InvalidDevice);
            }

            if (bufferSize < data.Count())
            {
                if (SetBufferSize(data.Count()) != 0)
                {
                    return -((int)ErrorCodes.MemholeDeviceKmallocFail);
                }
            }

            device.Write(data, 0, data.Count());

            return 0;
        }


    }
}
